### PR TITLE
EOS-14293: add version id initialization

### PR DESCRIPTION
**Pull request by evgeniy-brazhnikov**
Signed-off-by: evgeniy-brazhnikov <evgeniy.brazhnikov@seagate.com>

### CODE REVIEW

**Review by knrajnambiar76 — APPROVED** [review]

**Comment by evgeniy-brazhnikov** [comment]
http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/S3server/176/

**Comment by evgeniy-brazhnikov** [comment]
STs finished successfully on Dev VM too

File: server/s3_object_metadata.cc [hunk 1]
```diff
@@ -670,6 +670,10 @@ int S3ObjectMetadata::from_json(std::string content) {
   account_name = system_defined_attribute["Owner-Account"];
   account_id = system_defined_attribute["Owner-Account-id"];
   object_version_id = system_defined_attribute["x-amz-version-id"];
+
+  if (object_version_id.empty()) {
```
Comment (amitwac2608) [inline server/s3_object_metadata.cc:674]: when version_id could be empty ?

We saved the object metadata with empty version ID, in previous PUT object call  ? @knrajnambiar76, @DPG17  !!

File: server/s3_object_versioning_helper.cc [hunk 2]
```diff
@@ -54,7 +54,7 @@ std::string S3ObjectVersioingHelper::get_versionid_from_epoch_time(
 std::string S3ObjectVersioingHelper::generate_keyid_from_versionid(
     std::string versionid) {
   // add padding if requried;
-  versionid = versionid.append((3 - versionid.size() % 3) % 3, '=');
+  versionid = versionid.append((4 - (versionid.size() & 3)) & 3, '=');
```
Comment (amitwac2608) [inline server/s3_object_versioning_helper.cc:57]: @evgeniy-brazhnikov can you explain a little on why this change ?

File: server/s3_object_metadata.cc (hunk 1 above)
Comment (amitwac2608) [reply server/s3_object_metadata.cc:674]: yes, but we need to understand why it was empty ? Is there some issue with PUT request !!

File: server/s3_object_versioning_helper.cc (hunk 2 above)
Comment (evgeniy-brazhnikov) [reply server/s3_object_versioning_helper.cc:57]: x % 4 === x & 3 for each integer x.
So formula is (4 - (versionid.size() % 4)) %4
There was the error before.
Bitwise AND is much faster then division.

File: server/s3_object_metadata.cc (hunk 1 above)
Comment (DPG17) [reply server/s3_object_metadata.cc:674]: object_version_id, in this from_json() function, should never be empty.

Comment (DPG17) [reply server/s3_object_metadata.cc:674]: We need to investigate why object_version_id is empty.

File: server/s3_object_versioning_helper.cc (hunk 2 above)
Comment (amitwac2608) [reply server/s3_object_versioning_helper.cc:57]: ok.

**Review by DPG17 — CHANGES_REQUESTED** [review]
object_version_id, in this from_json() function, should never be empty.
@evgeniy-brazhnikov : you need to investigate why object_version_id is empty.

File: server/s3_object_metadata.cc (hunk 1 above)
Comment (evgeniy-brazhnikov) [reply server/s3_object_metadata.cc:674]: @knrajnambiar76 @DPG17 @amitwac2608 
M.b. something is incomplete while initializing multipart upload?

**Review by amitwac2608 — APPROVED** [review]

**Comment by evgeniy-brazhnikov** [comment]
http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/S3server/177/
Tests successfully finished on Dev VM also.

**Comment by evgeniy-brazhnikov** [comment]
Datta demanded to undo changes related to base64. Will do

File: server/s3_post_multipartobject_action.cc [hunk 3]
```diff
@@ -524,6 +524,8 @@ void S3PostMultipartObjectAction::save_upload_metadata() {
   object_multipart_metadata->set_content_type(request->get_content_type());
   object_multipart_metadata->set_part_index_oid(
       part_metadata->get_part_index_oid());
+  object_multipart_metadata->regenerate_version_id();
```
Comment (DPG17) [inline server/s3_post_multipartobject_action.cc:527]: This action class is about 'initiate multipart'. In this request, S3 server never creates object, but it is still not live to external world. When Post complete request is sent to S3 server, the object created here will be committed or will go live to external world. It is only during post complete request, the object is actually live, and the version is created during post complete action.

So, this regenerate_version_id() is not required here, as this is not final post complete action that makes the object live.

**Review by DPG17 — APPROVED** [review]

File: server/s3_post_multipartobject_action.cc (hunk 3 above)
Comment (DPG17) [reply server/s3_post_multipartobject_action.cc:527]: Ok, this regenerate_version_id() here will fix assert() issue that was observed during ST run.
Post complete action also calls regenerate_version_id() when object is actually becoming live.
